### PR TITLE
parquet reader: Support reading decimals from parquet `BYTE_ARRAY` type

### LINKED ISSUE #2159: read decimal data type from parquet file with binary physical type
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

From the definition of parquet spec format: https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L245

The decimal logical type can't be stored as `int32`,`int64`,`fixed binary`, and `binary`.

In the current arrow-parquet, we can't convert binary parquet data to decimal arrow data.

This is just compatible implementation for decimal in parquet.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #2159

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/buffer/converter.rs [hunk 1]
```diff
@@ -76,16 +76,6 @@ impl DecimalArrayConverter {
     pub fn new(precision: i32, scale: i32) -> Self {
         Self { precision, scale }
     }
-
-    fn from_bytes_to_i128(b: &[u8]) -> i128 {
```
Comment (liukun4515) [inline parquet/src/arrow/buffer/converter.rs:80]: move this method out and as a public method for this mod

File: parquet/src/arrow/schema.rs [hunk 2]
```diff
@@ -531,6 +531,32 @@ mod tests {
...
+
+        let parquet_schema = SchemaDescriptor::new(Arc::new(parquet_group_type));
+        let converted_arrow_schema =
+            parquet_to_arrow_schema(&parquet_schema, None).unwrap();
+
+        let arrow_fields = vec![
+            Field::new("decimal1", DataType::Decimal(4,2), false),
+            Field::new("decimal2", DataType::Decimal(12,2), false),
+            Field::new("decimal3", DataType::Decimal(30,2), false),
+            Field::new("decimal4", DataType::Decimal(33,2), false),
```
Comment (alamb) [inline parquet/src/arrow/schema.rs:555]: this would this have previously failed?

File: parquet/src/arrow/array_reader/builder.rs [hunk 3]
```diff
@@ -233,6 +225,18 @@ fn build_primitive_reader(
                 column_desc,
                 arrow_type,
             ),
+            Some(DataType::Decimal(precision, scale)) => {
+                // read decimal data from parquet binary physical type
+                let convert = DecimalByteArrayConvert::new(DecimalArrayConverter::new(precision as i32, scale as i32));
```
Comment (alamb) [inline parquet/src/arrow/array_reader/builder.rs:230]: Is this the core change?

 (namely to support reading arrow `DataType::Decimal` from parquet `BYTE_ARRAY` type where previously the code only supported `DataType::Decimal` in `FIXED_LEN_BYTE_ARRAY fields)?

**Review by alamb — APPROVED** [review]
This change makes sense to me -- thank you @liukun4515

**Comment by alamb** [comment]
Other reviewers might find the context on https://github.com/apache/arrow-datafusion/pull/2960/files#r928343239 helpful

File: parquet/src/arrow/array_reader/builder.rs [hunk 4]
```diff
@@ -233,6 +225,18 @@ fn build_primitive_reader(
...
+                Ok(Box::new(
+                    ComplexObjectArrayReader::<ByteArrayType,DecimalByteArrayConvert>::new(
```
Comment (viirya) [inline parquet/src/arrow/array_reader/builder.rs:232]: style?

```suggestion
                    ComplexObjectArrayReader::<ByteArrayType, DecimalByteArrayConvert>::new(
```

File: parquet/src/arrow/schema.rs (hunk 2 above)
Comment (viirya) [reply parquet/src/arrow/schema.rs:555]: Yea, this has previously failed as I just ran this test to confirm.

```
---- arrow::schema::tests::test_decimal_fields stdout ----
thread 'arrow::schema::tests::test_decimal_fields' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError("Unable to convert parquet BYTE_ARRAY logical type Some(Decimal { scale: 2, precision: 33 }) or converted type DECIMAL")', parquet/src/arrow/schema.rs:1734:60
```

**Review by viirya — APPROVED** [review]

File: parquet/src/arrow/array_reader/builder.rs (hunk 3 above)
Comment (liukun4515) [reply parquet/src/arrow/array_reader/builder.rs:230]: Yes.
previously, decimal just can be read from `int32,int64,byte_array`, but in the definition of parquet decimal can be read from `int32,int64,byte_array,fixed_length_byte_array`.

File: parquet/src/arrow/array_reader/builder.rs (hunk 4 above)
Comment (liukun4515) [reply parquet/src/arrow/array_reader/builder.rs:232]: The code has been formatted.

**Pull request merged**